### DISCUSSION TITLE
Using gettext with menu! section! macro

### DISCUSSION THREAD

**Discussion by MTRNord** [Relm4/Relm4#357] [Q&A] [answered]
Hi I currently have this menu:

```rust
menu! {
        primary_menu: {
            section! {
                "_Preferences" => PreferencesAction,
                "_Keyboard" => ShortcutsAction,
                "_About PixelRust GTK" => AboutAction,
            }
        }
    }
```

How do I make it that I can translate the item names?

**Comment by AaronErhardt** [comment, marked as answer]
I'm not sure how to tell GTK to use gettext, but the macro should roughly generate the following code for the first action:

```rust
gio::MenuItem::new(Some("_Preferences"), Some(&PreferencesAction::action_name()));
```

Before fixing or improving the macro we first have to figure out how to tell GTK to translate the strings in regular Rust code. This might also be a question for the [gtk-rs matrix channel](https://matrix.to/#/#rust:gnome.org).

**Reply by MTRNord** [reply]
> I'm not sure how to tell GTK to use gettext, but the macro should roughly generate the following code for the first action:

The way the gtk-rs template repos do it is using `gettext-rs` which provides a `gettext()` function. So it would happen purely in rust for the most part. See https://gitlab.gnome.org/World/Rust/gtk-rust-template/-/blob/master/src/main.rs#L21 for example

**Reply by AaronErhardt** [reply]
Ok, so the solution is to call the `gettext()` function. But how do we integrate this into the macro? Should we automatically translate texts if a certain feature flag (for example called "gettext") is enabled? Or do you have a better idea?

**Reply by MTRNord** [reply]
Hm as I am not sure if everyone uses gettext (there are more libs in rust to do this in the end) it may make sense to instead have a way to override the name in the macro? So, allowing you to set it using some extra syntax? Basically having some `#[name=gettext("abc")]` thing or something similar to allow setting it. Not sure what the best syntax is here.

**Reply by AaronErhardt** [reply]
Instead of the string literals, we could also allow any expression in the macro. For example, this would be possible then:

```rust
menu! {
        primary_menu: {
            section! {
                &gettext("_Preferences") => PreferencesAction,
                &gettext("_Keyboard") => ShortcutsAction,
                &gettext("_About PixelRust GTK") => AboutAction,
            }
        }
    }
```

**Reply by MTRNord** [reply]
That looks like a clean solution to me.

**Reply by MTRNord** [reply]
@AaronErhardt what's the best way going forward with this to make it possible? :)

**Reply by AaronErhardt** [reply]
Opening an issue would be a good idea because discussions tend to be overlooked easily. I think this feature can be added without too much effort, but maybe it's a bit harder if you haven't programmed proc-macros before. If you don't have the time or resources to implement this (we're happy to guide you), someone in our team will take care of it within the next couple days.
